[PRFC] Add backend support for Backstage events

timbonicus opened the pull request:
## Backstage Events

Backstage plugins and external systems connecting to Backstage would benefit from an event system. This allows reactive, push-style integrations rather than polling and synchronizing with the Backstage catalog, for example.

For the catalog in particular, #8219 describes a need for an API endpoint to expose these such events. This PRFC takes a different approach than the database table described there, instead adapting the `events-backend` concepts to provide a core `EventService` that any plugin can use.

The `EventBroker` and related types have been hoisted to `backend-common`, and `backend-app-api` now provides a default in-memory event service/broker. The events are published to subscribers in real-time; this **does not** provide any storage mechanism. I posit that storage would be achieved by replacing the in-memory event service with an external system (Pub/Sub, SQS, SNS, Kafka, etc).

For a backend using the [experimental backend system](https://backstage.io/docs/backend-system/), no change is needed; the core event service has a `defaultFactory` to create the in-memory event broker. Backends using the previous system must add an `EventBroker` to `makeCreateEnv` (and update the `PluginEnvironment`):

```diff
function makeCreateEnv(config: Config) {
  const root = getRootLogger();
  ...
  return (plugin: string): PluginEnvironment => {
    const logger = root.child({ type: 'plugin', plugin });
    const database = databaseManager.forPlugin(plugin);
    const cache = cacheManager.forPlugin(plugin);
    const scheduler = taskScheduler.forPlugin(plugin);
+   const eventBroker = new DefaultEventBroker(logger);

    return {
      logger,
      permissions,
      scheduler,
      identity,
+     eventBroker,
};
```

Services in the new backend system can add a dependency on `coreServices.events`; backends using the previous system can use the `eventBroker` added in the environment.

This is perhaps a prerequisite for a notification system; events are discussed in #10652 in relation to notifications. This branch aims to maintain compatability with the existing `events-backend` rather than modifying the event shape. A new `BackstageEvent` type is introduced that extends the generic `EventParams` to add an event type and `originatingEntityRef` as potential Backstage-related data of interest.

Backend events have the potential to become a firehose of information, and given an external event broker this also has the potential to rack up real costs. The cost consideration is skirted by the default in-memory event broker, but it's unclear if this default is suitable for production systems. It certainly is only applicable to Backstage instances that run in a single process (e.g. not horizontally scaled to multiple pods).

The review the PR, I suggest starting with the changes in `backend-common` and `catalog-backend`. Due to moving types from `events-node`, there are a great deal of import-only changes.

The event `type` for catalog events added in this PRFC are similar to catalog permission names. It's unclear if these concepts overlap enough to consider how they might fit together.

### Alternatives

#### Plugin extension points

I explored using [extension points](https://backstage.io/docs/backend-system/architecture/extension-points) instead, such that events might be opt-in to enable for each plugin. For example, a plugin could expose an `addEventListener` extension point and only send events when event listeners are registered:

```ts
backend.add(catalogPlugin());
backend.add(catalogModuleEventsPublisher());
```

The downside is that each plugin must implement an extension point and corresponding publishing module, and Backstage adopters must add event publishers for each plugin. It feels like events should be a bit more built-in and require little effort on the part of plugin developers.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

freben commented on packages/backend-common/api-report.md:
```diff
@@ -110,6 +110,19 @@ export class AzureUrlReader implements UrlReader {
 }
 
+// @public
+export type BackstageEvent<
```
I wonder if there's really any value in subscribing to a topic about _the entirety of Backstage_. I feel like the catalog is an entire product unto itself, which may affect the event topic design. Ping @pjungermann here of course, does that match how you think about proper granularity of these things?

My gut feeling tells me that we might choose to remove `BackstageEvent` and `BackstageEventPayload`, and instead add

```ts
export type CatalogEvent<
  TPayload extends CatalogEventPayload = CatalogEventPayload
> = EventParams<TPayload> & {
  topic: 'backstage.catalog',
};

export type CatalogEventPayload = {
  type: string;
};
```

to `@backstage/plugin-catalog-node`.

Then we could have more specific event subtypes there as well if we wanted to. If you want to use originating entity ref as a type-invariant routing parameter, then sure you could add that too but it's a bit of a large decision so if there's no strict need for it now, might want to hold off for a bit and see what patterns emerge as we add more types.

On that topic (ha!), would it even have been compelling to put that type in `@backstage/plugin-catalog-common` so that it can be used from the frontend as well when we have subscriptions there in the future? Which would mean that at least some of the types in this file could then be living in a new `packages/events` named `@backstage/events` - or possibly in `@backstage/plugin-events-common`, but haven't considered yet whether that might introduce any weird circular dependencies.

timbonicus replied:
I like these ideas! I removed `BackstageEvent` in favor of `CatalogEvent` in catalog-common in https://github.com/backstage/backstage/pull/17174/commits/66298c08baa1d99da2e328224762404af83a2405.

I'm unsure about an `events` package; some of the types like `EventBroker` don't seem sensible for the frontend to know about, so would it then just be `EventParams` that's common? Or just expose all the types to keep them together even if it might not make sense for the frontend?

freben reviewed (commented):
Loving the initiative! <3

I wish GitHub had a feature "mark all as viewed that only have changes to import statements". :)

freben reviewed (commented):
Finally got around to reading through this properly :)

Rugvip commented on plugins/catalog-common/package.json:
```diff
@@ -44,6 +44,7 @@
     "clean": "backstage-cli package clean"
   },
   "dependencies": {
+    "@backstage/backend-common": "workspace:^",
```
👍 to `@backstage/events`, although a little bit depending on exactly what we put there 😁. It might be the case that we should just add it to a different package, split it up into two separate packages like export from both `@backstage/core-plugin-api` and `@backstage/backend-plugin-api`, or create a new package but make it more generic.

pjungermann commented on .github/CODEOWNERS:
```diff
@@ -48,7 +48,6 @@ yarn.lock                                           @backstage/maintainers @back
 /plugins/events-backend-module-github               @backstage/maintainers @pjungermann
 /plugins/events-backend-module-gitlab               @backstage/maintainers @pjungermann
-/plugins/events-backend-test-utils                  @backstage/maintainers @pjungermann
 /plugins/events-node                                @backstage/maintainers @pjungermann
```
as the new `/packages/events` content is an extraction of parts of `/plugins/events-node`, you can put/keep me as one of the CODEOWNERS.

pjungermann commented on packages/backend-app-api/src/services/implementations/events/eventServiceFactory.ts:
```diff
@@ -14,16 +14,19 @@
-  setEventBroker(eventBroker: EventBroker): Promise<void>;
-}
+/** @public */
+export const eventServiceFactory = createServiceFactory({
```
Maybe, it should be called `eventsServiceFactory` instead as the service is called `events`

pjungermann commented on packages/backend-plugin-api/src/services/definitions/EventService.ts:
```diff
@@ -14,10 +14,7 @@
 
-export * from './testUtils';
+/** @public */
+export interface EventService extends EventBroker {}
```
```suggestion
export interface EventsService extends EventBroker {}
```

pjungermann commented on packages/backend-plugin-api/src/services/definitions/coreServices.ts:
```diff
@@ -58,6 +58,15 @@ export namespace coreServices {
+   *
+   * @public
+   */
+  export const events = createServiceRef<import('./EventService').EventService>(
```
I agree with @Rugvip. Let's keep it simpler for now and extend if needed. If plugins have a specific need for certain information, they could use metadata or even the payload itself. Furthermore, I could imagine that we end up with namespaces for events, etc.

pjungermann commented on packages/events/src/index.ts:
```diff
@@ -0,0 +1,22 @@
+  EventBroker,
+  EventParams,
+  EventPublisher,
+  EventSubscriber,
```
With the new setup, there is open question whether we still need the publisher and subscriber interfaces.

EventPublisher implementations could receive the `events` instance at the constructor, same for subscribers. There would be no need to expose them at their respective public API.

However, this could be rather a follow-up I guess as changes at implementations of these would make this PR much bigger.

pjungermann commented on plugins/catalog-backend-module-bitbucket-cloud/package.json:
```diff
@@ -51,6 +51,7 @@
     "@backstage/catalog-client": "workspace:^",
     "@backstage/catalog-model": "workspace:^",
     "@backstage/config": "workspace:^",
+    "@backstage/events": "workspace:^",
```
Do we still need `plugins-events-node` here?

pjungermann replied:
same for other packages

timbonicus replied:
We do since the `SubTopicEventRouter` is still in events-node 🤔

pjungermann commented on packages/create-app/templates/default-app/packages/backend/src/types.ts:
```diff
@@ -22,4 +23,5 @@ export type PluginEnvironment = {
   scheduler: PluginTaskScheduler;
   permissions: PermissionEvaluator;
   identity: IdentityApi;
+  eventBroker: EventBroker;
```
I wonder whether we should just call this `events` like the service 🤔

pjungermann commented on plugins/catalog-backend/api-report.md:
```diff
@@ -172,6 +173,7 @@ export type CatalogEnvironment = {
   config: Config;
   reader: UrlReader;
   permissions: PermissionEvaluator | PermissionAuthorizer;
+  eventBroker?: EventBroker;
```
similar to `PluginEnvironment`:
I wonder whether we should just call this `events` like the service 🤔

pjungermann commented on plugins/catalog-common/src/events.ts:
```diff
@@ -14,25 +14,24 @@
+export type CatalogEvent<
+  TPayload extends CatalogEventPayload = CatalogEventPayload,
+> = EventParams<TPayload> & {
+  topic: 'backstage.catalog';
```
Maybe we want to have a more specific event topic? Not sure.

pjungermann commented on plugins/events-backend-module-azure/package.json:
```diff
@@ -35,14 +35,15 @@
+    "@backstage/backend-common": "workspace:^",
     "@backstage/backend-plugin-api": "workspace:^",
+    "@backstage/events": "workspace:^",
     "@backstage/plugin-events-node": "workspace:^",
```
Is this package still needed?

Same for similar files.

pjungermann reviewed (commented):
Looks like a great step forward to me 🚀 Thanks!

I left a few questions and suggestions.

timbonicus commented:
Thanks for the reviews @freben @Rugvip and @pjungermann! I think I addressed most of the feedback. There are two outstanding, related issues around the same piece of code (the "mother of all queries" heretofore **MOAQ**).

The first is that I added `returning` to the MOAQ, which @freben pointed out won't work in MySQL. We're trying not to intentionally make life more difficult for eventual MySQL support, should it come, so this is undesirable.

The second is that, after discussion with @freben, I added `originatingEntityUid` to catalog events alongside `originatingEntityRef`, as both may be of interest. The MOAQ deleteWithEagerPruning is the one place I haven't added the `uid` to events yet (see [TODO](https://github.com/backstage/backstage/pull/17174/files#diff-b4c37de4ac6aec3a28e511640ddde1e13547c6c8a49bfee6898aeeec6d5c77e0R95)).

To solve both of these, I think we essentially need to do a select-then-delete in the MOAQ instead of a straight delete, so that we can pluck off the `entity_ref`s and `uid`s to return without a `returning` clause. This is a bit scary though, certainly for me who hasn't worked closely with this piece of code before. This also seems like a riskier change than the other additions here that are opt-in and fairly innocuous additions. I've run into things in the past like `WHERE IN` clauses having a [limit on the number of parameters](https://dba.stackexchange.com/questions/228695/is-there-any-limit-for-in-results-in-sql-server) supported, so I'm a bit paranoid about this kind of change to a fundamental function.

What do you think is the best way to proceed?

Rugvip commented:
@timbonicus yep deleting is tricky, the way I approached it in #10115 is to separate out deletion, plugins/catalog-backend/src/processing/DefaultCatalogProcessingEngine.ts. Honestly I think for sanity it may make sense to move over to something similar to that to avoid large queries 😅

timbonicus commented:
@Rugvip So https://github.com/backstage/backstage/pull/10115 essentially makes the `entity_id` in `final_entites` NULL (through foreign key onDelete) when the `refresh_state` row with the same `entity_id` is deleted by MOAQ, right? So rather than fiddle with MOAQ, we can just observe the results after the fact in the processing engine.

Makes sense to me, just want to make sure I understand correctly before taking a stab at this.

Rugvip commented:
@timbonicus yep iirc 😅, along with a cleanup job that takes care of the actual deletion. There it's done with a plain interval, but I think it makes sense to do through a scheduled task instead.

efenner-cambia commented:
bump

efenner-cambia commented:
bump

pjungermann commented:
@timbonicus Is there anything I can help you with to get this closed?

Can we extract some parts into separate PRs maybe?

E.g., exposing the `EventBroker` at `PluginEnvironment` was already merged.

timbonicus commented:
@pjungermann Absolutely, would love help getting this across the line. This was a hack project that I haven't been able to get back to yet.

There are two major things in this PRFC - moving the event types into core packages, and adding catalog events. I could see separating these into two PRs, as I think the first seems to be uncontroversial in the comments here so far.

The second bit of catalog events is currently failing a bunch of database checks and needs more work due to schema changes. That's a result of stealing work from #10115 which had the same failures. Isolating those changes away from the bazillion import statement changes from moving types would be nice, though.

pjungermann commented:
FYI: I will be on vacation myself, however, I plan to spend some time on it when I'm back.

benjdlambert commented:
@pjungermann have a good holiday!

timbonicus commented:
Closing for now as this is a bit stale, PR available for reference if anyone has time to hack on this 😄
